libvast/vast/system/simple_store.hpp

```diff
@@ -0,0 +1,93 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include <unordered_map>
+
+#include <caf/none.hpp>
+
+#include "vast/data.hpp"
+#include "vast/filesystem.hpp"
+#include "vast/load.hpp"
+#include "vast/save.hpp"
+
+#include "vast/system/key_value_store.hpp"
+
+namespace vast::system {
+
+template <class Key, class Value>
+struct simple_store_state {
+  using simple_store_actor = typename key_value_store_type<
+    Key, Value>::template stateful_pointer<simple_store_state>;
+  static inline const char* name = "data-store";
+
+  std::unordered_map<Key, Value> store;
+  path file;
+
+  caf::error init(simple_store_actor self, const path& dir) {
+    file = dir / "store";
+    if (auto err = vast::load(self->system(), file, store)) {
+      VAST_WARNING_ANON(name, "unable to load state file:", file);
+      return err;
+    }
+    return caf::none;
+  }
+
+  caf::error save(caf::actor_system& sys) {
```

Comment:
I would prefer storing a `caf::event_based_actor* self` pointer as member. Then we don't need to pass an argument here but instead can access `self->system()`.